File: solutions/typescript/0069.ts
function mySqrt(x: number): number {
    if (x < 2) return x;
    let left = 1, right = Math.floor(x / 2);
    while (left <= right) {
        const mid = Math.floor((left + right) / 2);
        const sq = mid * mid;
        if (sq === x) return mid;
        else if (sq < x) left = mid + 1;
        else right = mid - 1;
    }
    return right;
}

export default mySqrt;

File: solutions/rust/0069.rs
impl Solution {
    pub fn my_sqrt(x: i32) -> i32 {
        if x < 2 {
            return x;
        }
        let (mut left, mut right) = (1i64, (x / 2) as i64);
        while left <= right {
            let mid = (left + right) / 2;
            let sq = mid * mid;
            if sq == x as i64 {
                return mid as i32;
            } else if sq < x as i64 {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
        right as i32
    }
}
